Client Development/Client Scripts/Illumio Max Retry Count onChange.js

```diff
@@ -0,0 +1,10 @@
+function onChange(control, oldValue, newValue, isLoading, isTemplate) {
```

Comment:
Do we need to redefine this function for each onChange script? Could we instead define it as a helper function or does the it need to follow a specific callback spec?

Reply:
This is the onChange function given by ServiceNow's client scripts
Whenever we create a client script it automatically populates this structure, so now if we want to make a helper function we will have to make it into the Server side and call it from AJAX which will be a costly operation.

Reply:
So SNOW will call `onChange` with these 5 params, but we could then reference a client-side function rather than duplicating the logic, e.g.

```
function onChange(control, oldValue, newValue, isLoading, isTemplate) {
    if (isLoading || newValue === '') {
        return;
    }
    changeIntField(control, oldValue, newValue, isLoading, isTemplate, 'http_retry_interval_increment', 0, HTTP_RETRY_INTERVAL_MAX)
}

// ... in some helper script

function changeIntField(control, oldValue, newValue, isLoading, isTemplate, fieldName, minValue, maxValue) {
    newValue = newValue.replaceAll(',', '');
    var val = parseInt(newValue, 10);
    if (val < minValue || val > maxValue) {
        g_form.showFieldMsg(fieldName, `Invalid value. Please enter valid value between ${minValue}-${maxValue}`, 'error');
    }
}
```

Reply:
On the client side should we be using `gs.setProperty`/`getProperty` for these values?

Reply:
There is no direct way to do this, but there is a way by which we can achieve this

We need to make a UI script that has the code for validating the value which will look something like this (this is the structure that ServiceNow provides and i have added the validateInt method in it)

```
var x_illu2_illumio = x_illu2_illumio || {};

x_illu2_illumio.IllumioUIScript = (function() {
    "use strict";

    /* set your private variables and functions here. For example: 
    	var privateVar = 0; 
    	function private_function() {
    		return ++privateVar;
    	}
    */

    /* Share variables between multiple UI scripts by adding them to your scope object. For example: 
    	x_illu2_illumio.sharedVar = 0; 

     Then access them in your scripts the same way. For example: 
    	function get_shared() {
    		return x_illu2_illumio.sharedVar;
    	}
    */

    return {

        /* set your public API here. For example:
        		incrementAndReturnPrivateVar: function() {
        			return private_function();
        		},
        */
        validateInt: function(oldValue, newValue, fieldName, minValue, maxValue) {
            newValue = newValue.replaceAll(',', '');
            var val = parseInt(newValue, 10);
            if (val < minValue || val > maxValue) {
                g_form.showFieldMsg(fieldName, 'Invalid value. Please enter valid value between ' + minValue + '-' + maxValue, 'error');
            }

        },
        type: "IllumioUIScript"
    };
})();
```

Now then the client script will look something like this:

```
function onChange(control, oldValue, newValue, isLoading, isTemplate) {
    if (isLoading || newValue === '') {
        return;
    }
    ScriptLoader.getScripts(['x_illu2_illumio.IllumioUIScript.jsdbx'], function() {
        x_illu2_illumio.IllumioUIScript.validateInt(oldValue, newValue, 'http_retry_count', 0, 100);
    });
}
```

**Also a note that as it is ES5, it doesn't yet support the inline substitution of the variables in a string.**

Let me know if you would like to go ahead with this approach
I can try if we can access the constant file in the UI script or not, if you want to move ahead with this idea.

Reply:
Template literals are also an ES6+ thing? Yikes.

I see, would the ScriptLoader call have a performance impact? Seems as though all the options make code reuse challenging here...

Reply:
I don't suspect that the ScriptLoader would have much performance impact in this use-case, but going forward if we keep using this, it could cause a performance impact i think

Reply:
I was just going through the Network tab while checking the onChange, it looks like that it does an API call to load the UI script, but it was done only once, so it is loaded in the client side and then it is getting used directly

Reply:
Ok, that makes sense. I think having a UI script as a common/helper module would work well then.